codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -284,194 +316,542 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
+            }
+            allocHandle.incMessagesRead(1);
+            pipeline().fireChannelRead(msg);
+            if (numBytesToBeConsumed != 0) {
+                bytesConsumed(numBytesToBeConsumed);
+            }
+            return true;
+        }
+
+        private void reevaluateWritability() {
+            ChannelOutboundBuffer buffer = unsafe().outboundBuffer();
+            // If the buffer is not writable but should be writable, then write and flush a dummy object
+            // to trigger a writability change.
+            if (!buffer.isWritable() && buffer.totalPendingWriteBytes() < config.getWriteBufferHighWaterMark()) {
+                buffer.addMessage(REEVALUATE_WRITABILITY_MESSAGE, 1, voidPromise());
+                unsafe().flush();
+            }
+        }
+
+        private void doWrite(Object msg, ChannelPromise childPromise) {
+            final ChannelFuture future;
+            if (msg instanceof Http2StreamFrame) {
+                Http2StreamFrame frame = validateStreamFrame(msg);
+                if (!firstFrameWritten && !isStreamIdValid(stream().id())) {
+                    if (!(frame instanceof Http2HeadersFrame)) {
+                        throw new IllegalArgumentException("The first frame must be a headers frame. Was: "
+                                + frame.name());
+                    }
+                    childPromise.addListener(firstFrameWriteListener);
+                    firstFrameWritten = true;
+                }
+                frame.stream(stream());
+
+                assert !childPromise.isCancellable();
+                future = ctx.write(frame);
+            } else if (msg instanceof Http2GoAwayFrame) {
+                future = ctx.write(msg);
+            } else {
+                String msgStr = msg.toString();
+                ReferenceCountUtil.release(msg);
+                throw new IllegalArgumentException(
+                        "Message must be an Http2GoAwayFrame or Http2StreamFrame: " + msgStr);
+            }
+            // Notify the child promise once complete the write
+            future.addListener(new ChannelPromiseNotifier(childPromise));
         }
 
-        Http2StreamChannel channel() {
-            return childChannel;
+        private void bytesConsumed(final int bytes) {
+            ctx.write(new DefaultHttp2WindowUpdateFrame(bytes).stream(stream()));
+        }
+
+        private static Http2StreamFrame validateStreamFrame(Object msg) {
+            if (!(msg instanceof Http2StreamFrame)) {
+                String msgString = msg.toString();
+                ReferenceCountUtil.release(msg);
+                throw new IllegalArgumentException("Message must be a Http2StreamFrame: " + msgString);
+            }
+            Http2StreamFrame frame = (Http2StreamFrame) msg;
+            if (frame.stream() != null) {
+                String msgString = msg.toString();
+                ReferenceCountUtil.release(frame);
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)